File: src-python/core/services/settings_chat.py
"""设定模式 AI 对话服务。参见 D-0024、D-0029。

精简版上下文组装 + LLM tool calling 请求。
AI 只建议不执行：返回 tool_calls 原样交给前端。
"""

from __future__ import annotations

import logging
import re
from pathlib import Path
from typing import Any, Optional

from core.domain.settings_tools import get_tools_for_mode
from core.domain.tokenizer import count_tokens
from infra.llm.openai_compatible import OpenAICompatibleProvider

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# 对话历史截断
# ---------------------------------------------------------------------------

_MAX_HISTORY_MESSAGES = 10  # 5 轮 = 10 条 messages


def _truncate_history(messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """保留最近 5 轮（10 条），截断最早的。"""
    if len(messages) <= _MAX_HISTORY_MESSAGES:
        return list(messages)
    return list(messages[-_MAX_HISTORY_MESSAGES:])


# ---------------------------------------------------------------------------
# System Prompt
# ---------------------------------------------------------------------------

# System prompt 常量已迁移到 core/prompts/zh.py 和 en.py，通过 get_prompts(language) 获取。


# ---------------------------------------------------------------------------
# Fandom DNA 摘要提取
# ---------------------------------------------------------------------------

_DNA_SECTION_RE = re.compile(
    r"^##\s*(核心本质|核心特质|Core Essence|Core Traits)",
    re.MULTILINE,
)


def _extract_dna_summary(content: str, max_chars: int = 1500) -> str:
    """从角色设定文件中提取核心段落摘要。

    优先提取 ## 核心本质 / ## 核心特质 段落，
    找不到则取前 max_chars 字符。
    每角色最多约 500 token（~1500 中文字符）。
    """
    match = _DNA_SECTION_RE.search(content)
    if match:
        start = match.start()
        # 找下一个 ## 标题或文件末尾
        next_heading = content.find("\n## ", start + 1)
        section = content[start:next_heading] if next_heading != -1 else content[start:]
        if len(section) > max_chars:
            section = section[:max_chars] + "…"
        return section.strip()

    # fallback：取前 max_chars 字符
    if len(content) > max_chars:
        return content[:max_chars] + "…"
    return content.strip()


# ---------------------------------------------------------------------------
# 上下文组装
# ---------------------------------------------------------------------------

def build_settings_context(
    mode: str,
    base_path: str,
    fandom_path: Optional[str],
    messages: list[dict[str, Any]],
    language: str = "zh",
) -> list[dict[str, Any]]:
    """组装设定模式的 messages 数组。

    不走写作模式的 P0-P5 六层组装。
    """
    from core.prompts import get_prompts
    P = get_prompts(language)

    system_parts: list[str] = []

    if mode == "au":
        # 读取 project 信息
        au_name, fandom_name = _load_au_meta(base_path)
        system_parts.append(
            P.SETTINGS_AU_SYSTEM_PROMPT.format(au_name=au_name, fandom_name=fandom_name)
        )

        # Fandom DNA 摘要
        if fandom_path:
            dna_summary = _load_fandom_dna_summaries(fandom_path)
            if dna_summary:
                system_parts.append(f"{P.SETTINGS_FANDOM_DNA_HEADER}\n{dna_summary}")

        # AU 上下文
        au_context = _load_au_context(base_path, language=language)
        if au_context:
            system_parts.append(au_context)

    elif mode == "fandom":
        fandom_name = Path(base_path).name
        system_parts.append(
            P.SETTINGS_FANDOM_SYSTEM_PROMPT.format(fandom_name=fandom_name)
        )

        # Fandom 现有设定文件全文
        fandom_files = _load_settings_files(
            Path(base_path),
            [("core_characters", P.SETTINGS_LABEL_CORE_CHARACTERS), ("core_worldbuilding", P.SETTINGS_LABEL_CORE_WORLDBUILDING)],
            language=language,
        )
        if fandom_files:
            system_parts.append(f"{P.SETTINGS_CURRENT_FANDOM_FILES_HEADER}\n{fandom_files}")

    system_content = "\n\n".join(system_parts)

    # 截断对话历史
    truncated = _truncate_history(messages)

    result: list[dict[str, Any]] = [
        {"role": "system", "content": system_content},
    ]
    result.extend(truncated)

    return result


# ---------------------------------------------------------------------------
# 辅助：加载 AU 元数据
# ---------------------------------------------------------------------------

def _load_au_meta(au_path: str) -> tuple[str, str]:
    """从 project.yaml 读取 AU 名称和 Fandom 名称。"""
    import yaml

    project_yaml = Path(au_path) / "project.yaml"
    if not project_yaml.is_file():
        au_name = Path(au_path).name
        return au_name, "Unknown"

    try:
        raw = yaml.safe_load(project_yaml.read_text(encoding="utf-8")) or {}
    except Exception:
        return Path(au_path).name, "Unknown"

    return raw.get("name", Path(au_path).name), raw.get("fandom", "Unknown")


_SETTINGS_FILES_TOKEN_LIMIT = 30000


def _load_settings_files(
    base_dir: Path,
    categories: list[tuple[str, str]],
    language: str = "zh",
) -> str:
    """读取指定目录下的 .md 设定文件全文。

    Args:
        base_dir: 根目录（AU 或 Fandom 路径）。
        categories: [(子目录名, 标签)] 列表，如 [("characters", "角色设定"), ("worldbuilding", "世界观")]。

    Returns:
        格式化的设定文件全文。超量时截断低 importance 角色。
    """
    entries: list[tuple[str, str, str]] = []  # (label, filename, content)
    total_chars = 0

    for subdir, label in categories:
        dir_path = base_dir / subdir
        if not dir_path.is_dir():
            continue
        for f in sorted(dir_path.iterdir()):
            if f.is_file() and f.suffix == ".md":
                try:
                    content = f.read_text(encoding="utf-8")
                    entries.append((label, f.name, content))
                    total_chars += len(content)
                except Exception:
                    continue

    if not entries:
        return ""

    # 超量保护：>30000 token 时截断（tokenizer 估算 1 token ≈ 0.67 字符）
    if total_chars > _SETTINGS_FILES_TOKEN_LIMIT // 1.5:
        logger.warning(
            "设定文件总量 %d 字符，超过阈值，截断低 importance 角色",
            total_chars,
        )
        entries = _truncate_low_importance(entries, language=language)

    parts: list[str] = []
    for label, filename, content in entries:
        parts.append(f"[{label}] {filename}:\n{content}")

    return "\n\n".join(parts)


def _truncate_low_importance(
    entries: list[tuple[str, str, str]],
    language: str = "zh",
) -> list[tuple[str, str, str]]:
    """截断低 importance 角色：只保留 frontmatter + ## 核心限制 段落。"""
    result: list[tuple[str, str, str]] = []
    for label, filename, content in entries:
        if _has_low_importance(content):
            truncated = _extract_frontmatter_and_core(content, language=language)
            result.append((label, filename, truncated))
        else:
            result.append((label, filename, content))
    return result


def _has_low_importance(content: str) -> bool:
    """检查 YAML frontmatter 中 importance 是否为 low（仅检查 frontmatter）。"""
    if not content.startswith("---"):
        return False
    try:
        import yaml
        fm_parts = content.split("---", 2)
        if len(fm_parts) < 3:
            return False
        fm = yaml.safe_load(fm_parts[1]) or {}
        return fm.get("importance") == "low"
    except Exception:
        return False


def _extract_frontmatter_and_core(content: str, language: str = "zh") -> str:
    """提取 frontmatter + ## 核心限制 段落。"""
    from core.prompts import get_prompts
    P = get_prompts(language)

    parts: list[str] = []

    # frontmatter
    if content.startswith("---"):
        fm_parts = content.split("---", 2)
        if len(fm_parts) >= 3:
            parts.append(f"---{fm_parts[1]}---")
            remaining = fm_parts[2]
        else:
            remaining = content
    else:
        remaining = content

    # ## 核心限制 段落
    core_match = re.search(r"^## (?:核心限制|Core Constraints).*?(?=\n## |\Z)", remaining, re.MULTILINE | re.DOTALL)
    if core_match:
        parts.append(core_match.group().strip())

    if not parts:
        return content[:500] + P.SETTINGS_TRUNCATED_SUFFIX

    return "\n\n".join(parts) + P.SETTINGS_TRUNCATED_FULL_SUFFIX


def _load_fandom_dna_summaries(fandom_path: str) -> str:
    """读取 Fandom core_characters/ 下所有角色的 DNA 摘要。"""
    chars_dir = Path(fandom_path) / "core_characters"
    if not chars_dir.is_dir():
        return ""

    parts: list[str] = []
    for f in sorted(chars_dir.iterdir()):
        if f.is_file() and f.suffix == ".md":
            try:
                content = f.read_text(encoding="utf-8")
                summary = _extract_dna_summary(content)
                if summary:
                    parts.append(f"### {f.stem}\n{summary}")
            except Exception:
                continue

    return "\n\n".join(parts)


def _load_au_context(au_path: str, language: str = "zh") -> str:
    """加载 AU 级上下文：设定文件全文 + pinned_context + writing_style。"""
    import yaml
    from core.prompts import get_prompts
    P = get_prompts(language)

    parts: list[str] = []

    # 设定文件全文（characters/ + worldbuilding/）
    files_text = _load_settings_files(
        Path(au_path),
        [("characters", P.SETTINGS_LABEL_CHARACTERS), ("worldbuilding", P.SETTINGS_LABEL_WORLDBUILDING)],
        language=language,
    )
    if files_text:
        parts.append(f"{P.SETTINGS_CURRENT_AU_FILES_HEADER}\n{files_text}")

    # pinned_context + writing_style 从 project.yaml 读取
    project_yaml = Path(au_path) / "project.yaml"
    if project_yaml.is_file():
        try:
            raw = yaml.safe_load(project_yaml.read_text(encoding="utf-8")) or {}
        except Exception:
            raw = {}

        pinned = raw.get("pinned_context", [])
        if pinned:
            lines = "\n".join(f"- {p}" for p in pinned)
            parts.append(f"{P.SETTINGS_CURRENT_PINNED_HEADER}\n{lines}")

        ws = raw.get("writing_style", {})
        if ws:
            ws_parts: list[str] = []
            if ws.get("perspective"):
                ws_parts.append(f"perspective: {ws['perspective']}")
            if ws.get("emotion_style"):
                ws_parts.append(f"emotion: {ws['emotion_style']}")
            if ws.get("custom_instructions"):
                ws_parts.append(f"custom: {ws['custom_instructions']}")
            if ws_parts:
                parts.append(f"{P.SETTINGS_CURRENT_STYLE_HEADER}\n{'  |  '.join(ws_parts)}")

    return "\n\n".join(parts)


# ---------------------------------------------------------------------------
# LLM 调用 + 响应解析
# ---------------------------------------------------------------------------

def call_settings_llm(
    assembled_messages: list[dict[str, Any]],
    tools: list[dict[str, Any]],
    llm_config: dict[str, str],
) -> dict[str, Any]:
    """调用 LLM（非流式 + tool calling），返回解析后的响应。

    Returns:
        {"content": str, "tool_calls": list[dict]}
    """
    provider = OpenAICompatibleProvider(
        api_base=llm_config.get("api_base", ""),
        api_key=llm_config.get("api_key", ""),
        model=llm_config.get("model", ""),
    )

    # 构造请求 body（直接使用 provider 内部方法复用重试逻辑）
    body: dict[str, Any] = {
        "model": provider._model,
        "messages": assembled_messages,
        "tools": tools,
        "tool_choice": "auto",
        "temperature": 0.7,
        "top_p": 0.95,
        "max_tokens": 4096,
        "stream": False,
    }

    data = provider._request_with_retry(body)

    # 解析响应
    content = ""
    tool_calls: list[dict[str, Any]] = []

    choices = data.get("choices", [])
    if choices:
        msg = choices[0].get("message", {})
        content = msg.get("content", "") or ""
        raw_tool_calls = msg.get("tool_calls") or []
        for tc in raw_tool_calls:
            tool_calls.append({
                "id": tc.get("id", ""),
                "type": tc.get("type", "function"),
                "function": {
                    "name": tc.get("function", {}).get("name", ""),
                    "arguments": tc.get("function", {}).get("arguments", "{}"),
                },
            })

    return {"content": content, "tool_calls": tool_calls}

File: src-engine/services/settings_chat.ts
// Copyright (c) 2026 FicForge Contributors
// Licensed under the GNU Affero General Public License v3.0.

/**
 * 设定模式 AI 对话服务。参见 D-0024、D-0029。
 * 精简版上下文组装 + LLM tool calling 请求。
 * AI 只建议不执行：返回 tool_calls 原样交给前端。
 */

import { get_tools_for_mode } from "../domain/settings_tools.js";
import { getPrompts } from "../prompts/index.js";
import type { PlatformAdapter } from "../platform/adapter.js";
import type { LLMProvider, Message, ToolCall } from "../llm/provider.js";
import { LLMError } from "../llm/provider.js";
import { hasLogger, getLogger } from "../logger/index.js";
import { joinPath } from "../repositories/implementations/file_utils.js";
import yaml from "js-yaml";

// ---------------------------------------------------------------------------
// 对话历史截断
// ---------------------------------------------------------------------------

const MAX_HISTORY_MESSAGES = 10;

function truncateHistory(messages: Message[]): Message[] {
  if (messages.length <= MAX_HISTORY_MESSAGES) return [...messages];
  return messages.slice(-MAX_HISTORY_MESSAGES);
}

// ---------------------------------------------------------------------------
// Fandom DNA 摘要提取
// ---------------------------------------------------------------------------

const DNA_SECTION_RE = /^##\s*(核心本质|核心特质|Core Essence|Core Traits)/m;

function extractDnaSummary(content: string, maxChars = 1500): string {
  const match = DNA_SECTION_RE.exec(content);
  if (match) {
    const start = match.index;
    const nextHeading = content.indexOf("\n## ", start + 1);
    let section = nextHeading !== -1 ? content.slice(start, nextHeading) : content.slice(start);
    if (section.length > maxChars) section = section.slice(0, maxChars) + "…";
    return section.trim();
  }
  if (content.length > maxChars) return content.slice(0, maxChars) + "…";
  return content.trim();
}

// ---------------------------------------------------------------------------
// 上下文组装
// ---------------------------------------------------------------------------

export interface SettingsChatParams {
  mode: "au" | "fandom";
  base_path: string;
  fandom_path?: string | null;
  messages: Message[];
  language?: string;
  adapter: PlatformAdapter;
}

export async function build_settings_context(params: SettingsChatParams): Promise<Message[]> {
  const { mode, base_path, fandom_path = null, messages, language = "zh", adapter } = params;
  const P = getPrompts(language as "zh" | "en");

  const systemParts: string[] = [];

  if (mode === "au") {
    const [auName, fandomName] = await loadAuMeta(base_path, adapter);
    systemParts.push(
      P.SETTINGS_AU_SYSTEM_PROMPT
        .replace("{au_name}", auName)
        .replace("{fandom_name}", fandomName),
    );

    if (fandom_path) {
      const dnaSummary = await loadFandomDnaSummaries(fandom_path, adapter);
      if (dnaSummary) {
        systemParts.push(`${P.SETTINGS_FANDOM_DNA_HEADER}\n${dnaSummary}`);
      }
    }

    const auContext = await loadAuContext(base_path, language, adapter);
    if (auContext) systemParts.push(auContext);
  } else {
    const fandomName = base_path.split("/").pop() ?? base_path;
    systemParts.push(
      P.SETTINGS_FANDOM_SYSTEM_PROMPT.replace("{fandom_name}", fandomName),
    );

    const fandomFiles = await loadSettingsFiles(base_path, [
      ["core_characters", P.SETTINGS_LABEL_CORE_CHARACTERS],
      ["core_worldbuilding", P.SETTINGS_LABEL_CORE_WORLDBUILDING],
    ], language, adapter);
    if (fandomFiles) {
      systemParts.push(`${P.SETTINGS_CURRENT_FANDOM_FILES_HEADER}\n${fandomFiles}`);
    }
  }

  const systemContent = systemParts.join("\n\n");
  const truncated = truncateHistory(messages);

  return [
    { role: "system", content: systemContent },
    ...truncated,
  ];
}

// ---------------------------------------------------------------------------
// LLM 调用
// ---------------------------------------------------------------------------

export interface SettingsChatResult {
  content: string;
  tool_calls: ToolCall[];
}

export async function call_settings_llm(
  assembled_messages: Message[],
  mode: "au" | "fandom",
  llm_provider: LLMProvider,
): Promise<SettingsChatResult> {
  const tools = get_tools_for_mode(mode);

  // 先尝试带 tool calling 请求；若模型/提供商不兼容则降级为纯文本。
  try {
    const response = await llm_provider.generate({
      messages: assembled_messages,
      max_tokens: 4096,
      temperature: 0.7,
      top_p: 0.95,
      tools: tools as unknown as import("../llm/provider.js").ToolDefinition[],
      tool_choice: "auto",
    });

    return {
      content: response.content,
      tool_calls: response.tool_calls ?? [],
    };
  } catch (err) {
    // 已明确分类的错误（上下文超限、内容过滤、非 400）原样抛出
    if (!(err instanceof LLMError) || err.status_code !== 400) throw err;
    if (err.error_code === "context_length_exceeded" || err.error_code === "content_filtered") throw err;

    // 400 且无明确分类 → 大概率是 tool calling 格式/数量不兼容，去掉 tools 重试
    if (hasLogger()) getLogger().warn("settings_chat", "tool calling 400, retrying without tools", { mode, error: err.message });
    const response = await llm_provider.generate({
      messages: assembled_messages,
      max_tokens: 4096,
      temperature: 0.7,
      top_p: 0.95,
    });

    return {
      content: response.content,
      tool_calls: [],
    };
  }
}

// ---------------------------------------------------------------------------
// 辅助：加载 AU 元数据
// ---------------------------------------------------------------------------

async function loadAuMeta(auPath: string, adapter: PlatformAdapter): Promise<[string, string]> {
  const projectPath = joinPath(auPath, "project.yaml");
  const exists = await adapter.exists(projectPath);
  if (!exists) {
    return [auPath.split("/").pop() ?? auPath, "Unknown"];
  }
  try {
    const text = await adapter.readFile(projectPath);
    const raw = (yaml.load(text) ?? {}) as Record<string, unknown>;
    return [
      (raw.name as string) ?? auPath.split("/").pop() ?? auPath,
      (raw.fandom as string) ?? "Unknown",
    ];
  } catch {
    return [auPath.split("/").pop() ?? auPath, "Unknown"];
  }
}

const SETTINGS_FILES_TOKEN_LIMIT = 30000;

async function loadSettingsFiles(
  baseDir: string,
  categories: [string, string][],
  language: string,
  adapter: PlatformAdapter,
): Promise<string> {
  const entries: [string, string, string][] = []; // [label, filename, content]
  let totalChars = 0;

  for (const [subdir, label] of categories) {
    const dirPath = joinPath(baseDir, subdir);
    const exists = await adapter.exists(dirPath);
    if (!exists) continue;

    const files = await adapter.listDir(dirPath);
    for (const f of files.sort()) {
      if (!f.endsWith(".md")) continue;
      try {
        const content = await adapter.readFile(joinPath(dirPath, f));
        entries.push([label, f, content]);
        totalChars += content.length;
      } catch {
        continue;
      }
    }
  }

  if (entries.length === 0) return "";

  // 超量保护
  let finalEntries = entries;
  if (totalChars > SETTINGS_FILES_TOKEN_LIMIT / 1.5) {
    finalEntries = truncateLowImportance(entries, language);
  }

  return finalEntries.map(([label, filename, content]) => `[${label}] ${filename}:\n${content}`).join("\n\n");
}

function truncateLowImportance(
  entries: [string, string, string][],
  language: string,
): [string, string, string][] {
  return entries.map(([label, filename, content]) => {
    if (hasLowImportance(content)) {
      return [label, filename, extractFrontmatterAndCore(content, language)];
    }
    return [label, filename, content];
  });
}

function hasLowImportance(content: string): boolean {
  if (!content.startsWith("---")) return false;
  try {
    const parts = content.split("---", 3);
    if (parts.length < 3) return false;
    const fm = yaml.load(parts[1]) as Record<string, unknown> | null;
    return fm?.importance === "low";
  } catch {
    return false;
  }
}

function extractFrontmatterAndCore(content: string, language: string): string {
  const P = getPrompts(language as "zh" | "en");
  const parts: string[] = [];

  let remaining = content;
  if (content.startsWith("---")) {
    const fmParts = content.split("---", 3);
    if (fmParts.length >= 3) {
      parts.push(`---${fmParts[1]}---`);
      remaining = fmParts[2];
    }
  }

  // 提取 ## 核心限制 段落（到下一个 ## 标题或文件末尾）
  // 先找标题起始位置
  const headingMatch = remaining.match(/^## (?:核心限制|Core Constraints)/m);
  if (headingMatch && headingMatch.index !== undefined) {
    const start = headingMatch.index;
    // 找下一个 ## 标题
    const nextHeading = remaining.indexOf("\n## ", start + 1);
    const section = nextHeading !== -1 ? remaining.slice(start, nextHeading) : remaining.slice(start);
    parts.push(section.trim());
  }

  if (parts.length === 0) {
    return content.slice(0, 500) + P.SETTINGS_TRUNCATED_SUFFIX;
  }

  return parts.join("\n\n") + P.SETTINGS_TRUNCATED_FULL_SUFFIX;
}

async function loadFandomDnaSummaries(fandomPath: string, adapter: PlatformAdapter): Promise<string> {
  const charsDir = joinPath(fandomPath, "core_characters");
  const exists = await adapter.exists(charsDir);
  if (!exists) return "";

  const files = await adapter.listDir(charsDir);
  const parts: string[] = [];
  for (const f of files.sort()) {
    if (!f.endsWith(".md")) continue;
    try {
      const content = await adapter.readFile(joinPath(charsDir, f));
      const summary = extractDnaSummary(content);
      if (summary) {
        const stem = f.replace(/\.md$/, "");
        parts.push(`### ${stem}\n${summary}`);
      }
    } catch {
      continue;
    }
  }
  return parts.join("\n\n");
}

async function loadAuContext(auPath: string, language: string, adapter: PlatformAdapter): Promise<string> {
  const P = getPrompts(language as "zh" | "en");
  const parts: string[] = [];

  const filesText = await loadSettingsFiles(
    auPath,
    [["characters", P.SETTINGS_LABEL_CHARACTERS], ["worldbuilding", P.SETTINGS_LABEL_WORLDBUILDING]],
    language, adapter,
  );
  if (filesText) {
    parts.push(`${P.SETTINGS_CURRENT_AU_FILES_HEADER}\n${filesText}`);
  }

  // pinned_context + writing_style 从 project.yaml 读取
  const projectPath = joinPath(auPath, "project.yaml");
  const exists = await adapter.exists(projectPath);
  if (exists) {
    try {
      const text = await adapter.readFile(projectPath);
      const raw = (yaml.load(text) ?? {}) as Record<string, unknown>;

      const pinned = (raw.pinned_context as string[]) ?? [];
      if (pinned.length > 0) {
        const lines = pinned.map((p) => `- ${p}`).join("\n");
        parts.push(`${P.SETTINGS_CURRENT_PINNED_HEADER}\n${lines}`);
      }

      const ws = (raw.writing_style ?? {}) as Record<string, unknown>;
      if (ws) {
        const wsParts: string[] = [];
        if (ws.perspective) wsParts.push(`perspective: ${ws.perspective}`);
        if (ws.emotion_style) wsParts.push(`emotion: ${ws.emotion_style}`);
        if (ws.custom_instructions) wsParts.push(`custom: ${ws.custom_instructions}`);
        if (wsParts.length > 0) {
          parts.push(`${P.SETTINGS_CURRENT_STYLE_HEADER}\n${wsParts.join("  |  ")}`);
        }
      }
    } catch {
      // ignore
    }
  }

  return parts.join("\n\n");
}
